File: 1461-check-if-a-string-contains-all-binary-codes-of-size-k/1461-check-if-a-string-contains-all-binary-codes-of-size-k.py
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        
        hashy = set()
        left = 0

        for right in range(len(s)):

            if right - left + 1 == k:

                subStr = s[left:right + 1]
                hashy.add(subStr)
                left += 1
        
        if len(hashy) == 2 ** k:
            return True
        
        return False

File: 1461-check-if-a-string-contains-all-binary-codes-of-size-k/1461-check-if-a-string-contains-all-binary-codes-of-size-k.cpp
class Solution {
public:
    bool hasAllCodes(string s, int k) 
    {
        int n = (int)s.size();

        //Base case
        if (n < k)
        {
            return false;
        }

        //Fast way to compute (2 ^ n)
        int need = 1 << k;

        unordered_set<std::string> res;
        int left = 0;
        std::string window;

        for (int right = 0; right < n; right++)
        {
            window.push_back(s[right]);

            if (window.size() > k) 
            {
                window.erase(window.begin());  //O(k)
            }

            if (window.size() == k) {
                res.insert(window);

                //Check if we got it...
                if (res.size() == need)
                {
                    return true;
                }
            }
        }

        return false;
    }
};